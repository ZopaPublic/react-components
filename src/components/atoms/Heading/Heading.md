### Summary

Whenever you want to render a heading on any UI at Zopa, use the `<Heading />` component.

Apply margin/padding by [adding spacing](/#/Content?id=spacing).

### Examples

- Variations

```tsx
import { Fragment } from 'react';
import { colors, Heading, Text } from '@zopauk/react-components';

<Fragment>
  <Heading as="h1" size="display">
    Display Heading
  </Heading>
  <Text color={colors.alert}>❗️ Only to be used for marketing purposes.</Text>
  <hr style={{ margin: '30px 0' }} />
  <Heading as="h1">Heading level 1</Heading>
  <Heading as="h2">Heading level 2</Heading>
  <Heading as="h3">Heading level 3</Heading>
  <Heading as="h4">Heading level 4</Heading>
  <Heading as="h5">Heading level 5</Heading>
  <Heading as="h6">Heading level 6</Heading>
</Fragment>;
```

- Sizing

```jsx
import { Fragment } from 'react';
import { Heading } from '@zopauk/react-components';

<Fragment>
  <Heading as="h2">As a h2 I render at 28px by default</Heading>
  <Heading as="h4">As a h4 I render at 20px by default</Heading>
  <Heading as="h2" size="h4">
    I render with a h2 tag but small
  </Heading>
  <Heading as="span" size="h3">
    I render as a {'<span>'} and big
  </Heading>
</Fragment>;
```

- Colors

```tsx { "props": { "style": { "background": "linear-gradient(180deg, rgba(0,217,197,1) 0%, rgba(255,255,255,1) 100%)", "border": "none" } } }
import { Fragment } from 'react';
import { Heading, colors } from '@zopauk/react-components';

<Fragment>
  <Heading as="h4" color={colors.white}>
    White
  </Heading>
  <Heading as="h4" color={colors.grey}>
    Near Dark
  </Heading>
  <Heading as="h4" color={colors.greyDarkest} className="mb-0">
    Dark
  </Heading>
</Fragment>;
```

- Alignment

```tsx
import { Fragment } from 'react';
import { Heading } from '@zopauk/react-components';

<Fragment>
  <div style={{ textAlign: 'right' }}>
    <Heading as="h4" className="mb-7">
      Inherits from his parent by default
    </Heading>
  </div>
  <Heading as="h4" align="left">
    Left aligned
  </Heading>
  <Heading as="h4" align="center">
    Center
  </Heading>
  <Heading as="h4" align="right" className="mb-0">
    Right aligned
  </Heading>
</Fragment>;
```

- Using forwardedAs:
  If Heading is styled it should use the 'forwardedAs' prop in place of as.
  If Heading is not styled the 'as' prop should be used.

```jsx
import { Fragment } from 'react';
import styled from 'styled-components';
import { Heading } from '@zopauk/react-components';

const StyledHeading = styled(Heading)``;

<Fragment>
  <Heading as="h4">Heading uses as prop</Heading>
  <StyledHeading forwardedAs="h4">Styled heading uses forwardedAs prop</StyledHeading>
</Fragment>;
```
