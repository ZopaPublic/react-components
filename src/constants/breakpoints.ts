export const breakpoints = {
  desktop: 1280,
  phone: 600,
  tablet: 720,
} as const;

export type DeviceSizes = keyof typeof breakpoints;
