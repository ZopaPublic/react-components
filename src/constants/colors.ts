const brandColors = {
  brand: '#009E8E',
  brandDark: '#063737',
  brandLight: '#ECFCF9',
  brandDecorative: '#00D9C5',
} as const;

const actionColors = {
  action: 'linear-gradient(90deg, #4F5AD8 0%, #3B46C4 100%)',
  actionPlain: '#3B46C4',
  actionDark: '#21247F',
  actionLight: '#E9EAFA',
} as const;

const neutralColors = {
  white: '#FFFFFF',
  greyLightest: '#F7F7F7',
  greyLighter: '#EFEFEF',
  greyLight: '#D4D7D9',
  grey: '#818F9B',
  greyDark: '#4A545E',
  greyDarkest: '#2C3236',
} as const;

const notificationColors = {
  alert: '#FF4539',
  alertDark: '#940700',
  alertLight: '#FFDAD8',
  warning: '#FF9F0A',
  warningDark: '#704300',
  warningLight: '#FFECCE',
  success: '#3EBC64',
  successDark: '#17592B',
  successLight: '#DDFDE5',
} as const;

export type Colors = typeof brandColors & typeof actionColors & typeof neutralColors & typeof notificationColors;

const colors: Colors = {
  ...brandColors,
  ...actionColors,
  ...neutralColors,
  ...notificationColors,
};

export {
  colors,
  // Exported only for documentation purposes in `<Colors />`
  brandColors,
  actionColors,
  neutralColors,
  notificationColors,
};
