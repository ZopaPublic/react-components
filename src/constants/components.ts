export const navbarHeight = 120;
export const mobileNavbarHeight = 64;
