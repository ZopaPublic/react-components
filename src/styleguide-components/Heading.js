import React from 'react';
import cx from 'clsx';
import Styled from 'rsg-components/Styled';

const styles = ({ color, fontSize }) => ({
  heading: {
    margin: '1em 0',
    color: color.base,
    fontWeight: 900,
  },
  heading1: {
    fontSize: fontSize.h1,
  },
  heading2: {
    fontSize: fontSize.h2,
  },
  heading3: {
    fontSize: fontSize.h3,
  },
  heading4: {
    fontSize: fontSize.h4,
  },
  heading5: {
    fontSize: fontSize.h5,
    fontWeight: 'bold',
  },
  heading6: {
    fontSize: fontSize.h6,
    fontStyle: 'italic',
  },
});

function HeadingRenderer({ classes, level, children, ...props }) {
  const Tag = `h${level}`;
  const headingClasses = cx(classes.heading, classes[`heading${level}`]);
  return (
    <Tag {...props} className={headingClasses}>
      {children}
    </Tag>
  );
}

export default Styled(styles)(HeadingRenderer);
