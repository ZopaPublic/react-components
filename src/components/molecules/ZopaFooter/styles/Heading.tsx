import styled from 'styled-components';
import Heading from '../../../atoms/Heading/Heading';

export const FooterHeading = styled(Heading)`
  text-transform: uppercase;
`;
