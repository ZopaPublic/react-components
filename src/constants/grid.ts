import { breakpoints } from './breakpoints';

export type GridBreakpoints = 'xs' | 's' | 'm' | 'l' | 'xl';
export type GridWidths = 's' | 'm' | 'l' | 'xl';

interface Grid {
  breakpoints: Record<GridBreakpoints, number>;
  cols: number;
  gutter: number;
  width: Record<GridWidths, number>;
}

const grid: Grid = {
  breakpoints: {
    xs: 0,
    xl: breakpoints.largeDesktop,
    l: breakpoints.desktop,
    m: breakpoints.tablet,
    s: breakpoints.phone,
  },
  cols: 12,
  gutter: 12,
  width: {
    l: 960,
    m: 720,
    s: 540,
    xl: 1224,
  },
};

export default grid;
