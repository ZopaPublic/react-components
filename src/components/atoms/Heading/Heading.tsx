import { grid } from '../../../constants';
import styled, { css } from 'styled-components';
import React, { HTMLAttributes } from 'react';
import { Colors } from '../../../constants/colors';
import { useThemeContext, AppThemeProps, AppTheme } from '../../styles/Theme';

export type HeadingTags = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'span';

export interface OptionalHeadingProps extends HTMLAttributes<HTMLHeadingElement> {
  /**
   * Override the default size assigned to the rendered HTML tag.
   * @default `as`
   */
  size?: keyof AppTheme['typography']['heading']['sizes'];
  /**
   * Accepts a subset of the Zopa brand colors by default. Same as the ones accepted by `<Text />`.
   * @default `colors.greyDarkest`
   */
  color?: Colors['white'] | Colors['grey'] | Colors['greyDarkest'] | string;
  /**
   * Where the rendered text should be aligned to.
   * @default 'inherit'
   */
  align?: 'inherit' | 'left' | 'right' | 'center';
}

interface HeadingPropsWithAs extends OptionalHeadingProps {
  children?: React.ReactNode;
  /**
   * The HTML5 tag you want to render your heading, it's used to determine the size of the heading as well.
   * @default 'required if 'forwardedAs' is not provided.'
   */
  as: HeadingTags;
}

interface HeadingPropsWithForwardedAs extends OptionalHeadingProps {
  /**
   * The HTML5 tag you want to render your heading, it's used to determine the size of the heading as well.
   * To be used if Component has been wrapped in styled components.
   * @default 'required if 'as' is not provided.'
   */
  forwardedAs: HeadingTags;
}

export type HeadingProps = HeadingPropsWithAs | HeadingPropsWithForwardedAs;

const StyledHeading = styled.h1<HeadingPropsWithAs & AppThemeProps>`
  ${({ as, size, theme }) => {
    if (as) {
      const tag = size || (as === 'span' ? 'h4' : as);
      return css`
        font-size: ${theme.typography.heading.sizes[tag]};
        line-height: ${theme.typography.lineHeight[tag]};
        letter-spacing: ${theme.typography.letterSpacingMap[tag]};
      `;
    }
  }};

  margin: 0;
  color: ${({ color, theme }) => color || theme.typography.text.color};
  display: ${({ as }) => (as === 'span' ? 'block' : undefined)};
  text-align: ${({ align = 'inherit' }) => align};
  font-family: ${({ theme }) => theme.typography.primary};
  font-weight: ${({ as, theme }) => theme.typography.weights[['h1', 'display'].includes(as) ? 'extraBold' : 'bold']};

  ${({ as, size, theme }) =>
    as === 'h1' &&
    size === 'display' &&
    css`
      @media screen and (max-width: ${grid.breakpoints.l}px) {
        font-size: ${theme.typography.heading.sizes['h2']};
        line-height: ${theme.typography.lineHeight['h2']};
        letter-spacing: ${theme.typography.letterSpacingMap['h2']};
        font-weight: ${theme.typography.weights['bold']};
      }
    `}
`;

export const Heading = (props: HeadingProps) => {
  const theme = useThemeContext();
  const headingProps = props as HeadingPropsWithAs;
  if (!headingProps?.as) {
    console.warn('Heading component is missing the as prop');
  }
  return <StyledHeading {...headingProps} theme={theme} />;
};

export default Heading;
