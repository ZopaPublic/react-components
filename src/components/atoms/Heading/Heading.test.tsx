import axe from '../../../../axe-helper';
import React from 'react';
import { render } from '@testing-library/react';
import { colors, typography } from '../../../constants';
import Heading from './Heading';
const {
  sizes: { heading: headingSizes },
} = typography;

describe('<Heading />', () => {
  it('renders without  a11y violations', async () => {
    const { container } = render(<Heading as="h1">Text</Heading>);
    const results = await axe(container.innerHTML);

    expect(results).toHaveNoViolations();
    expect(container).toMatchSnapshot();
  });

  it('renders display size as h2 at mobile', async () => {
    const { container } = render(
      <Heading as="h1" size="display">
        Text
      </Heading>,
    );

    expect(container).toMatchSnapshot();
  });

  it.each`
    tag
    ${'h1'}
    ${'h2'}
    ${'h3'}
    ${'h4'}
    ${'h5'}
    ${'h6'}
  `('it can render with a different HTML tag: $tag', ({ tag }) => {
    const { container } = render(<Heading as={tag}>Header</Heading>);

    expect(container.firstChild).toMatchSnapshot();
    expect(container.querySelector(tag)).not.toBe(null);
  });

  it('allows to control the size despite the rendered HTML tag', () => {
    const { container } = render(
      <Heading as="h1" size="h3">
        Header
      </Heading>,
    );

    expect(container.firstChild).toHaveStyleRule('font-size', headingSizes.h3);
  });

  it('renders with H4 size as default when rendering a `span`', () => {
    const { container } = render(<Heading as="span">Header</Heading>);

    expect(container.firstChild).toHaveStyleRule('font-size', headingSizes.h4);
  });

  it('allows to control the size when rendering as a `span`', () => {
    const { container } = render(
      <Heading as="span" size="h4">
        Header
      </Heading>,
    );

    expect(container.firstChild).toHaveStyleRule('font-size', headingSizes.h4);
  });

  it.each`
    hex                   | name
    ${colors.white}       | ${'White'}
    ${colors.greyDarkest} | ${'Dark'}
  `('can render in different colors: $name â€“ $hex', ({ hex }) => {
    const { container } = render(
      <Heading as="h2" color={hex}>
        Text
      </Heading>,
    );
    expect(container.firstChild).toHaveStyleRule('color', hex);
  });

  it('can render without margin bottom', () => {
    const { container } = render(<Heading as="h1">Header</Heading>);

    expect(container.firstChild).toHaveStyleRule('margin', '0');
  });

  it.each`
    direction
    ${'left'}
    ${'right'}
    ${'center'}
  `('can align at:  $direction', ({ direction }) => {
    const { container } = render(
      <Heading as="h4" align={direction}>
        Text
      </Heading>,
    );

    expect(container.firstChild).toHaveStyleRule('text-align', direction);
  });
});
