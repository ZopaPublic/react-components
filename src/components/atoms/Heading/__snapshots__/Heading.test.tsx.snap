// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Heading /> it can render with a different HTML tag: h1 1`] = `
.c0 {
  font-size: 46px;
  line-height: 54px;
  -webkit-letter-spacing: -1.25px;
  -moz-letter-spacing: -1.25px;
  -ms-letter-spacing: -1.25px;
  letter-spacing: -1.25px;
  margin: 0;
  color: #2C3236;
  text-align: inherit;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 800;
}

<h1
  class="c0"
>
  Header
</h1>
`;

exports[`<Heading /> it can render with a different HTML tag: h2 1`] = `
.c0 {
  font-size: 38px;
  line-height: 46px;
  -webkit-letter-spacing: -0.85px;
  -moz-letter-spacing: -0.85px;
  -ms-letter-spacing: -0.85px;
  letter-spacing: -0.85px;
  margin: 0;
  color: #2C3236;
  text-align: inherit;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 700;
}

<h2
  class="c0"
>
  Header
</h2>
`;

exports[`<Heading /> it can render with a different HTML tag: h3 1`] = `
.c0 {
  font-size: 28px;
  line-height: 36px;
  -webkit-letter-spacing: -0.45px;
  -moz-letter-spacing: -0.45px;
  -ms-letter-spacing: -0.45px;
  letter-spacing: -0.45px;
  margin: 0;
  color: #2C3236;
  text-align: inherit;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 700;
}

<h3
  class="c0"
>
  Header
</h3>
`;

exports[`<Heading /> it can render with a different HTML tag: h4 1`] = `
.c0 {
  font-size: 24px;
  line-height: 32px;
  -webkit-letter-spacing: -0.25px;
  -moz-letter-spacing: -0.25px;
  -ms-letter-spacing: -0.25px;
  letter-spacing: -0.25px;
  margin: 0;
  color: #2C3236;
  text-align: inherit;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 700;
}

<h4
  class="c0"
>
  Header
</h4>
`;

exports[`<Heading /> it can render with a different HTML tag: h5 1`] = `
.c0 {
  font-size: 18px;
  line-height: 26px;
  -webkit-letter-spacing: -0.02px;
  -moz-letter-spacing: -0.02px;
  -ms-letter-spacing: -0.02px;
  letter-spacing: -0.02px;
  margin: 0;
  color: #2C3236;
  text-align: inherit;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 700;
}

<h5
  class="c0"
>
  Header
</h5>
`;

exports[`<Heading /> it can render with a different HTML tag: h6 1`] = `
.c0 {
  font-size: 16px;
  line-height: 24px;
  -webkit-letter-spacing: -0.01px;
  -moz-letter-spacing: -0.01px;
  -ms-letter-spacing: -0.01px;
  letter-spacing: -0.01px;
  margin: 0;
  color: #2C3236;
  text-align: inherit;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 700;
}

<h6
  class="c0"
>
  Header
</h6>
`;

exports[`<Heading /> renders display size as h2 at mobile 1`] = `
.c0 {
  font-size: 68px;
  line-height: 76px;
  -webkit-letter-spacing: -2.86px;
  -moz-letter-spacing: -2.86px;
  -ms-letter-spacing: -2.86px;
  letter-spacing: -2.86px;
  margin: 0;
  color: #2C3236;
  text-align: inherit;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 800;
}

@media screen and (max-width:992px) {
  .c0 {
    font-size: 38px;
    line-height: 46px;
    -webkit-letter-spacing: -0.85px;
    -moz-letter-spacing: -0.85px;
    -ms-letter-spacing: -0.85px;
    letter-spacing: -0.85px;
    font-weight: 700;
  }
}

<div>
  <h1
    class="c0"
  >
    Text
  </h1>
</div>
`;

exports[`<Heading /> renders without  a11y violations 1`] = `
.c0 {
  font-size: 46px;
  line-height: 54px;
  -webkit-letter-spacing: -1.25px;
  -moz-letter-spacing: -1.25px;
  -ms-letter-spacing: -1.25px;
  letter-spacing: -1.25px;
  margin: 0;
  color: #2C3236;
  text-align: inherit;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 800;
}

<div>
  <h1
    class="c0"
  >
    Text
  </h1>
</div>
`;
